File: prince-of-versions/src/main/java/co/infinum/princeofversions/NetworkLoader.java
package co.infinum.princeofversions;

import android.util.Base64;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;

/**
 * Represents a concrete loader that load resource from network using provided URL.
 */
public class NetworkLoader implements Loader {

    /**
     * Default request timeout in seconds.
     */
    private static final int DEFAULT_NETWORK_TIMEOUT_SECONDS = 60;

    private static final int MILLISECONDS_IN_SECOND = 1000;

    /**
     * Url representing the resource.
     */
    private String url;

    /**
     * Custom network timeout in seconds.
     */
    private int networkTimeoutMilliseconds = DEFAULT_NETWORK_TIMEOUT_SECONDS * MILLISECONDS_IN_SECOND;

    /**
     * Basic authentication username.
     */
    private String username;

    /**
     * Basic authentication password.
     */
    private String password;

    /**
     * Creates a new network loader using provided url.
     *
     * @param url Resource locator.
     */
    public NetworkLoader(String url) {
        this(url, DEFAULT_NETWORK_TIMEOUT_SECONDS);
    }

    /**
     * Creates a new network loader using url and custom network timeout.
     *
     * @param url                   Resource locator.
     * @param networkTimeoutSeconds Custom network timeout.
     */
    public NetworkLoader(String url, int networkTimeoutSeconds) {
        this(url, null, null, networkTimeoutSeconds);
    }

    /**
     * Creates a new network loader using url, default network timeout and basic authentication parameters.
     *
     * @param url      Resource locator.
     * @param username Basic authentication username.
     * @param password Basic authentication password.
     */
    public NetworkLoader(String url, String username, String password) {
        this(url, username, password, DEFAULT_NETWORK_TIMEOUT_SECONDS);
    }

    /**
     * Creates a new network loader using url, custom network timeout and basic authentication parameters.
     *
     * @param url                   Resource locator.
     * @param username              Basic authentication username.
     * @param password              Basic authentication password.
     * @param networkTimeoutSeconds Custom network timeout.
     */
    public NetworkLoader(String url, String username, String password, int networkTimeoutSeconds) {
        this.url = url;
        this.username = username;
        this.password = password;
        this.networkTimeoutMilliseconds = networkTimeoutSeconds * MILLISECONDS_IN_SECOND;
    }

    @Override
    public String load() throws IOException {

        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        try {
            if (username != null && password != null) {
                String credentials = username + ":" + password;
                String basicAuth = "Basic " + Base64.encodeToString(credentials.getBytes(Charset.forName("UTF-8")), Base64.NO_WRAP);
                conn.setRequestProperty("Authorization", basicAuth);
            }
            conn.setConnectTimeout(networkTimeoutMilliseconds);
            conn.setReadTimeout(networkTimeoutMilliseconds);
            InputStream response = conn.getInputStream();
            return StreamIo.toString(response);
        } finally {
            close(conn);
        }
    }

    /**
     * Closing http connection.
     *
     * @param conn Http connection.
     */
    protected void close(HttpURLConnection conn) {
        if (conn != null) {
            try {
                conn.disconnect();
            } catch (Exception ignorable) { // NOPMD
                // ignorable exception
            }
        }
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/NetworkLoader.kt
package co.infinum.princeofversions

import android.util.Base64
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.Charset

/**
 * Represents a concrete loader that loads a resource from the network using a provided URL.
 *
 * @param url The URL representing the resource locator.
 * @param username Optional username for Basic authentication.
 * @param password Optional password for Basic authentication.
 * @param networkTimeoutSeconds The network timeout in seconds.
 */
class NetworkLoader (
    private val url: String,
    private val username: String? = null,
    private val password: String? = null,
    networkTimeoutSeconds: Int = DEFAULT_NETWORK_TIMEOUT_SECONDS
) : Loader {

    companion object {
        /**
         * Default request timeout in seconds.
         */
        private const val DEFAULT_NETWORK_TIMEOUT_SECONDS = 60
        private const val MILLISECONDS_IN_SECOND = 1000
    }

    /**
     * Custom network timeout in milliseconds.
     */
    private val networkTimeoutMilliseconds = networkTimeoutSeconds * MILLISECONDS_IN_SECOND

    @Throws(IOException::class)
    override fun load(): String {
        val connection = URL(url).openConnection() as HttpURLConnection
        try {
            // Apply Basic Authentication if credentials are provided
            if (username != null && password != null) {
                val credentials = "$username:$password"
                val basicAuth = "Basic ${Base64.encodeToString(credentials.toByteArray(Charsets.UTF_8), Base64.NO_WRAP)}"
                connection.setRequestProperty("Authorization", basicAuth)
            }

            connection.connectTimeout = networkTimeoutMilliseconds
            connection.readTimeout = networkTimeoutMilliseconds

            // The 'use' block automatically handles closing the input stream
            return connection.inputStream.use { responseStream ->
                StreamIo.toString(responseStream)
            }
        } finally {
            connection.disconnect()
        }
    }
}
